src/base/misc/dos2linux.c

```diff
@@ -556,28 +571,30 @@ static void check_read_pagefault(dosaddr_t addr,
       handler(addr, 4, 0, 0);
     /* only add writable pages to the cache! */
     if (!dpmi_write_access(addr))
-      return;
+      return dosaddr_to_unixaddr(addr);
   }
-  set_unprotected_page(addr);
+  return set_unprotected_dosaddr_to_unixaddr(addr);
 }
 
 uint8_t do_read_byte(dosaddr_t addr, sim_pagefault_handler_t handler)
 {
-  if (mem_likely_protected(addr, 1)) {
+  void *uaddr = unprotected_dosaddr_to_unixaddr(addr, 1);
+  if (!uaddr) {
     /* use vga_write_access instead of vga_read_access here to avoid adding
        read-only addresses to the cache */
     if (vga_write_access(addr))
       return vga_read(addr);
     if (config.mmio_tracing && mmio_check(addr))
       return mmio_trace_byte(addr, READ_BYTE(addr), MMIO_READ);
-    check_read_pagefault(addr, handler);
+    uaddr = check_read_pagefault(addr, handler);
```

Comment:
Would it be possible to split that
uaddr things into a separate patch?
I never looked into your caching code
before, so I can't grok large patches in
that area. :)

Reply:
The present patch is a bit smaller, so is it ok as is. I'll have a look how to still split it up, but it's not obvious to me.